### ISSUE TITLE
Building with variants in conda_build_config.yaml fails with packages with "-" in their name

### ISSUE THREAD

**Issue by agrossfield** [conda/conda-build#5477]
### Checklist

- [X] I added a descriptive title
- [X] I searched open reports and couldn't find a duplicate

### What happened?

Hi,

I support a package that compiles against boost, and I'm trying to support multiple versions of the library. To specify both boost and libboost-devel at the same versions, I'm using the zip_keys functionality. 

My conda_build_config.yaml looks like:
```
boost:
  - 1.84
  - 1.85
  - 1.86
libboost: 
  - 1.84
  - 1.85
  - 1.86
libboost-devel: 
  - 1.84
  - 1.85
  - 1.86
zip_keys:
  - boost
  - libboost
  - libboost-devel
pin_run_as_build:
  boost: x.x
```

When conda-forge runs the tests on the pull request, everything works fine on linux, but on OSX I get the following error:
```
 conda-build ./recipe -m ./.ci_support/osx_64_python3.11.____cpython.yaml --suppress-variables --clobber-file ./.ci_support/clobber_osx_64_python3.11.____cpython.yaml --extra-meta flow_run_id=azure_20240904.5.1 remote_url=*** sha=c545751cf7a124951e3a868a578a8c41b811e0d1
WARNING: No numpy version specified in conda_build_config.yaml.  Falling back to default numpy value of 1.26
Traceback (most recent call last):
  File "/Users/runner/miniforge3/bin/conda-build", line 11, in <module>
    sys.exit(execute())
             ^^^^^^^^^
  File "/Users/runner/miniforge3/lib/python3.12/site-packages/conda_build/cli/main_build.py", line 589, in execute
    api.build(
  File "/Users/runner/miniforge3/lib/python3.12/site-packages/conda_build/api.py", line 209, in build
    return build_tree(
           ^^^^^^^^^^^
  File "/Users/runner/miniforge3/lib/python3.12/site-packages/conda_build/build.py", line 3686, in build_tree
    metadata_tuples = render_recipe(
                      ^^^^^^^^^^^^^^
  File "/Users/runner/miniforge3/lib/python3.12/site-packages/conda_build/render.py", line 990, in render_recipe
    variants = get_package_variants(m, variants=variants)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/miniforge3/lib/python3.12/site-packages/conda_build/variants.py", line 697, in get_package_variants
    combined_spec, specs = get_package_combined_spec(
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/miniforge3/lib/python3.12/site-packages/conda_build/variants.py", line 661, in get_package_combined_spec
    validate_spec(f, spec)
  File "/Users/runner/miniforge3/lib/python3.12/site-packages/conda_build/variants.py", line 200, in validate_spec
    raise ValueError(
ValueError: Variant configuration errors in /Users/runner/work/1/s/conda_build_config.yaml:
  libboost-devel key contains an invalid character '-'
```

Basically, if you try to zipper on a package that has a "-" in the name, it fails, but only on OS X. This seems like a bug.

I'm going to try to work around it, but I thought you'd want to know.

Thanks,

Alan

### Conda Info

```shell
irrelevant, the bug occurs in the conda-forge azure testing pipeline
```

### Conda Config

```shell
irrelevant, the bug occurs in the conda-forge azure testing pipeline
```

### Conda list

```shell
irrelevant, the bug occurs in the conda-forge azure testing pipeline
```

**Comment by agrossfield** [comment]
Follow-up: The bug isn't specific to using zip_keys. The following conda_build_config.yaml also triggers the same error:
```libboost-devel: 
  - 1.84
  - 1.85
  - 1.86
pin_run_as_build:
  libboost-boost: x.x
```

The error is:
```

+ [[ -f LICENSE.txt ]]
+ cp LICENSE.txt recipe/recipe-scripts-license.txt
+ [[ 0 == 1 ]]
+ conda-build ./recipe -m ./.ci_support/osx_64_python3.9.____cpython.yaml --suppress-variables --clobber-file ./.ci_support/clobber_osx_64_python3.9.____cpython.yaml --extra-meta flow_run_id=azure_20240904.6.1 remote_url=*** sha=36b62a3eb44c1dc16bb4a1eb696ec95cd9888797
WARNING: No numpy version specified in conda_build_config.yaml.  Falling back to default numpy value of 1.26
Traceback (most recent call last):
  File "/Users/runner/miniforge3/bin/conda-build", line 11, in <module>
    sys.exit(execute())
             ^^^^^^^^^
  File "/Users/runner/miniforge3/lib/python3.12/site-packages/conda_build/cli/main_build.py", line 589, in execute
    api.build(
  File "/Users/runner/miniforge3/lib/python3.12/site-packages/conda_build/api.py", line 209, in build
    return build_tree(
           ^^^^^^^^^^^
  File "/Users/runner/miniforge3/lib/python3.12/site-packages/conda_build/build.py", line 3686, in build_tree
    metadata_tuples = render_recipe(
                      ^^^^^^^^^^^^^^
  File "/Users/runner/miniforge3/lib/python3.12/site-packages/conda_build/render.py", line 990, in render_recipe
    variants = get_package_variants(m, variants=variants)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/miniforge3/lib/python3.12/site-packages/conda_build/variants.py", line 697, in get_package_variants
    combined_spec, specs = get_package_combined_spec(
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/miniforge3/lib/python3.12/site-packages/conda_build/variants.py", line 661, in get_package_combined_spec
    validate_spec(f, spec)
  File "/Users/runner/miniforge3/lib/python3.12/site-packages/conda_build/variants.py", line 200, in validate_spec
    raise ValueError(
ValueError: Variant configuration errors in /Users/runner/work/1/s/conda_build_config.yaml:
  libboost-devel key contains an invalid character '-'

```

**Comment by agrossfield** [comment]
One more detail I forgot to mention: This bug occurs in every OS X version that conda-forge tries to build with, so it happens with python 3.8. 3.10, 3.11, and 3.12. The same python versions are fine under linux.

**Comment by isuruf** [comment]
You have to use `_` instead of `-` in `recipe/conda_build_config.yaml`. This is documented in https://docs.conda.io/projects/conda-build/en/latest/resources/variants.html

**Comment by agrossfield** [comment]
No, it’s not documented there  — none of their examples involve pinning packages that have dashes in the name, and they don’t mention that case. I don’t have the option to change the name of boost-devel to something else, and I find it hard to believe that the correct solution is to misspell the name of the package (eg refer to boost-devel as boost_devel). I’ll try it, but if that’s the solution it needs to be documented in giant screaming letters (especially since boost incorrectly is featured as an example, when that package is almost never needed to run, and thus doesn’t have compatibility issues).

Dr. Alan Grossfield
Department of Biochemistry and Biophysics
University of Rochester Medical Center

On Oct 23, 2024, at 11:54 AM, Isuru Fernando ***@***.***> wrote:

﻿

You have to use _ instead of - in recipe/conda_build_config.yaml. This is documented in https://docs.conda.io/projects/conda-build/en/latest/resources/variants.html<https://urldefense.com/v3/__https://docs.conda.io/projects/conda-build/en/latest/resources/variants.html__;!!MkQhnT5yHH0MPvOc00c!WaaKYkNkMNYytmGAjbQPmGJYK-ZfBy-MEvysst_GdAMRms6JzzJoxj3GVIQqTyEwy2K3WH2bnWpMtRwPIfl6gfFjuKgJEOJ41EbfEw$>

**Comment by isuruf** [comment]
> none of their examples involve pinning packages that have dashes in the name, and they don’t mention that case. 

There's an example with `r-base`.

**Comment by agrossfield** [comment]
Ok, but my second point stands: the substitution is made without comment or explanation (the words “dash” or “hyphen” don’t appear anywhere on the page), during an explanation of something else. Even if I’d noticed it, I would have assumed it was a typo (especially given that several of the examples, eg the ones requiring specific versions of boost) don’t make much sense.

How on earth was anyone supposed to guess that the correct solution is to misspell the name of the package?

________________________________
Dr. Alan Grossfield
Dept of Biochemistry and Biophysics
University of Rochester Medical Center
Phone: 585 276 4193
http://membrane.urmc.rochester.edu
https://orcid.org/0000-0002-5877-2789
Pronouns: He/his



From: Isuru Fernando ***@***.***>
Date: Wednesday, October 23, 2024 at 12:17 PM
To: conda/conda-build ***@***.***>
Cc: Grossfield, Alan ***@***.***>, Author ***@***.***>
Subject: [EXT] Re: [conda/conda-build] Building with variants in conda_build_config.yaml fails with packages with "-" in their name (Issue #5477)

none of their examples involve pinning packages that have dashes in the name, and they don’t mention that case.

There's an example with r-base.

**Comment by isuruf** [comment]
Can you send a PR to the docs with your suggested wording?

Also, please tone down your language.

**Comment by agrossfield** [comment]
I apologize for my tone, but I’m a bit frustrated by yours – your two messages basically said I was an idiot for not knowing the solution, and I feel justified in pointing out that no one would have guessed the solution you implied was obvious.

I’ll try to send a PR when I have a few minutes, but I’m not a conda-build developer. The obvious place to put in the bare minimum message would be right below the r-base example:
------
Please note that, unlike the other files used to control conda’s build process, conda_build_config.yaml directly converts package names into variables. As such, this means that packages whose names contain characters that are illegal in variable names (e.g. “-“) will cause the build to fail. To work around this, one must substitute an underscore for the dashes in a package name. Thus, in the above example, the package is referenced as “r-base” in the meta.yaml but “r_base” in conda_build_config.yaml.
----

The other place that would make sense would be to modify the place where it says:
------
Variant input is ultimately a dictionary. These dictionaries are mostly very flat. Keys are made directly available in Jinja2 templates. As a result, keys in the dictionary (and in files read into dictionaries) must be valid jinja2 variable names (no - characters allowed). If you need to reference packages that have dashes in their names in conda_build_config.yaml, substitute underscore characters for the dashes.
------

________________________________
Dr. Alan Grossfield
Dept of Biochemistry and Biophysics
University of Rochester Medical Center
Phone: 585 276 4193
http://membrane.urmc.rochester.edu
https://orcid.org/0000-0002-5877-2789
Pronouns: He/his



From: Isuru Fernando ***@***.***>
Date: Wednesday, October 23, 2024 at 12:31 PM
To: conda/conda-build ***@***.***>
Cc: Grossfield, Alan ***@***.***>, Author ***@***.***>
Subject: [EXT] Re: [conda/conda-build] Building with variants in conda_build_config.yaml fails with packages with "-" in their name (Issue #5477)

Can you send a PR to the docs with your suggested wording?

Also, please tone down your language.

**Comment by agrossfield** [comment]
One other thing: nothing explains why the old way (referencing libboost-devel instead of libboost_devel) worked correctly on conda-forge for the linux builds but not the OSX builds.

________________________________
Dr. Alan Grossfield
Dept of Biochemistry and Biophysics
University of Rochester Medical Center
Phone: 585 276 4193
http://membrane.urmc.rochester.edu
https://orcid.org/0000-0002-5877-2789
Pronouns: He/his



From: Grossfield, Alan ***@***.***>
Date: Wednesday, October 23, 2024 at 12:45 PM
To: conda/conda-build ***@***.***>, conda/conda-build ***@***.***>
Cc: Author ***@***.***>
Subject: Re: [EXT] Re: [conda/conda-build] Building with variants in conda_build_config.yaml fails with packages with "-" in their name (Issue #5477)
I apologize for my tone, but I’m a bit frustrated by yours – your two messages basically said I was an idiot for not knowing the solution, and I feel justified in pointing out that no one would have guessed the solution you implied was obvious.

I’ll try to send a PR when I have a few minutes, but I’m not a conda-build developer. The obvious place to put in the bare minimum message would be right below the r-base example:
------
Please note that, unlike the other files used to control conda’s build process, conda_build_config.yaml directly converts package names into variables. As such, this means that packages whose names contain characters that are illegal in variable names (e.g. “-“) will cause the build to fail. To work around this, one must substitute an underscore for the dashes in a package name. Thus, in the above example, the package is referenced as “r-base” in the meta.yaml but “r_base” in conda_build_config.yaml.
----

The other place that would make sense would be to modify the place where it says:
------
Variant input is ultimately a dictionary. These dictionaries are mostly very flat. Keys are made directly available in Jinja2 templates. As a result, keys in the dictionary (and in files read into dictionaries) must be valid jinja2 variable names (no - characters allowed). If you need to reference packages that have dashes in their names in conda_build_config.yaml, substitute underscore characters for the dashes.
------

________________________________
Dr. Alan Grossfield
Dept of Biochemistry and Biophysics
University of Rochester Medical Center
Phone: 585 276 4193
http://membrane.urmc.rochester.edu
https://orcid.org/0000-0002-5877-2789
Pronouns: He/his



From: Isuru Fernando ***@***.***>
Date: Wednesday, October 23, 2024 at 12:31 PM
To: conda/conda-build ***@***.***>
Cc: Grossfield, Alan ***@***.***>, Author ***@***.***>
Subject: [EXT] Re: [conda/conda-build] Building with variants in conda_build_config.yaml fails with packages with "-" in their name (Issue #5477)

Can you send a PR to the docs with your suggested wording?

Also, please tone down your language.

**Comment by isuruf** [comment]
> I apologize for my tone

Thanks

> but I’m a bit frustrated by yours – your two messages basically said I was an idiot for not knowing the solution, and I feel justified in pointing out that no one would have guessed the solution you implied was obvious.

Sure. You should point it out without the tone. If you felt like my messages were insinuating that, you should mention that instead of you doing the same.

>  I’ll try to send a PR when I have a few minutes, but I’m not a conda-build developer.

Please do. You don't have to be a conda-build developer to send a PR. Someone will look at your PR and make suggestions if it needs any.
